core/src/main/java/com/yahoo/oak/FirstLevelHashArray.java

```diff
@@ -17,56 +17,76 @@
+     * num of pointers are going to point to another chunk and so on.
      * Other creation patterns (like all the hash array entries referencing the same chunk) are possible.
      * Majority of the parameters are needed for chunk creation.
-     *
-     * @param msbForFirstLevelHash number of most significant bits to be used to calculate the
+     *  @param msbForFirstLevelHash number of most significant bits to be used to calculate the
      *                             index in the hash array. It also affects directly the size of the
      *                             array. It is going to be incorporated inside hashIndexCodec and
      *                             transferred to chunks.
+     * @param lsbForSecondLevelHash
+     * @param multipleReferenceNum number of the references to reference the same chunk,
+     *                             before switching to next chunk
      *
-     * IMPORTANT: not thread safe, should be done upon raising of OakHashMap
      */
-    FirstLevelHashArray(int msbForFirstLevelHash, AtomicInteger externalSize, MemoryManager vMM,
-        MemoryManager kMM, OakComparator<K> comparator, OakSerializer<K> keySerializer,
-        OakSerializer<V> valueSerializer) {
+    FirstLevelHashArray(int msbForFirstLevelHash, int lsbForSecondLevelHash, AtomicInteger externalSize,
+        MemoryManager vMM, MemoryManager kMM, OakComparator<K> comparator,
+        OakSerializer<K> keySerializer, OakSerializer<V> valueSerializer, int multipleReferenceNum) {
 
         // the key hash (int) separation between MSB for first level ans LSB for second level,
         // to be used by hash array and all chunks, until resize
         this.hashIndexCodec =
             new UnionCodec(UnionCodec.AUTO_CALCULATE_BIT_SIZE, // the size of the first, as these are LSBs
-                msbForFirstLevelHash, Integer.SIZE); // the second (MSB) will be msbForFirstLevelHash
+                msbForFirstLevelHash, // the second (MSB) will be msbForFirstLevelHash
+                1, // and disregard the first bit as it is the sign bit and always zero
+                Integer.SIZE);
 
         // the size of the hash array is defined by number of MSBs to be used from key hash
         int arraySize = (int) (Math.pow(2, msbForFirstLevelHash));
         this.chunks = new AtomicReferenceArray<>(arraySize);
         this.msbForFirstLevelHash = msbForFirstLevelHash;
 
+        int currentSameRefer = multipleReferenceNum;
+        HashChunk<K, V> c = null;
+        UnionCodec hashIndexCodecForChunk =
+            (lsbForSecondLevelHash == InternalOakHash.USE_DEFAULT_FIRST_TO_SECOND_BITS_PARTITION)
+                ? this.hashIndexCodec :
+                new UnionCodec(lsbForSecondLevelHash, // the size of the first, as these are LSBs
+                    UnionCodec.AUTO_CALCULATE_BIT_SIZE, Integer.SIZE); // the second (MSB) will be auto-calculated
```

Comment:
Can you add an explanation of how these two codecs interact?
I think a small bit diagram will help.

Reply:
added a comment, please let me know if it is clear